cpp/src/arrow/util/concatenate.cc

```diff
@@ -0,0 +1,335 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/concatenate.h"
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/memory_pool.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/visibility.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+static inline ArrayData SliceData(const ArrayData& data, int64_t offset, int64_t length) {
+  DCHECK_LE(offset, data.length);
+  length = std::min(data.length - offset, length);
+  offset += data.offset;
+
+  auto copy = data;
+  copy.length = length;
+  copy.offset = offset;
+  copy.null_count = data.null_count != 0 ? kUnknownNullCount : 0;
+  return copy;
+}
+
+struct ConcatenateImpl {
+  ConcatenateImpl(const std::vector<ArrayData>& in, MemoryPool* pool)
+      : in_(in),
+        in_size_(static_cast<int>(in.size())),
+        lengths_(in.size()),
+        offsets_(in.size()),
+        pool_(pool) {
+    out_.type = in[0].type;
+    for (int i = 0; i != in_size_; ++i) {
+      lengths_[i] = in[i].length;
+      offsets_[i] = in[i].offset;
+      out_.length += lengths_[i];
+      if (out_.null_count == kUnknownNullCount || in[i].null_count == kUnknownNullCount) {
+        out_.null_count = kUnknownNullCount;
+        continue;
+      }
+      out_.null_count += in[i].null_count;
+    }
+  }
+
+  /// offset, length pair for representing a range of a buffer or array
+  struct range {
+    int32_t offset, length;
+
+    range() : offset(-1), length(0) {}
+    range(int32_t o, int32_t l) : offset(o), length(l) {}
+    void widen_to_include(int32_t index) {
+      if (length == 0) {
+        offset = index;
+        length = 1;
+        return;
+      }
+      auto end = std::max(offset + length, index + 1);
+      length = end - offset;
+      offset = std::min(index, offset);
+    }
+  };
+
+  Status Visit(const NullType&) { return Status::OK(); }
+
+  Status Visit(const BooleanType&) {
+    std::shared_ptr<Buffer> values_buffer;
+    RETURN_NOT_OK(ConcatenateBitmaps(1, &values_buffer));
+    out_.buffers.push_back(values_buffer);
+    return Status::OK();
+  }
+
+  // handle numbers, decimal128, fixed_size_binary
+  Status Visit(const FixedWidthType& fixed) {
```

Comment:
I think it's worth thinning the body of such functions by separating the core algorithm and the intermediary data construction , e.g.

```c++
std::vector<std::shared_ptr<Buffer>> FixedValueSlicedBuffers(const FixedWidthType& fixed) {
  ...
}

Status Visit(const FixedWidthType& fixed) {
  // The constructor (or top-most Visit) ensure that &out_.buffers[1] is already set
  return arrow::Concatenate(ValueSlicedBuffers(fixed), pool_, &out_.buffers[1]);
}
```